### ISSUE TITLE
Kodi 18.4 crashes with high bitrate TrueHD+Atmos audio stream

### ISSUE THREAD

**Issue by HomerJau** [xbmc/xbmc#16704]
### Describe the bug
When playing either an MKV or MKA file containing high bitrate (16,000kps) audio stream with HDMI pass-through Kodi 18.4 crashes: after a couple of seconds and audio begins to stutter.

I can replicate on Intel NUC8i3 (LibreElec 9.2), Odroid N2 (CoreElec latest nightly) and Windows 10 Pro (Kod 18.4 x64).

I have been playing Atmos music blu-ray rips to MKA and MKV successfully (actually I have had very occasional audio dropouts). These Atmos+TrueHD streams are usually around 7000kbps max. I have just converted the new The Beatles - Abbey Road 50th Ann. Blu-ray Audio and the Atmos stream is 16,000kbps. These files crash Kodi.

The same file playback without issue in Foobar2000.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

These files should bitstream via HDMI to my Denon AVR and play in full Atmos without crashing Kodi and without any audio dropouts.

## Actual Behavior

After selecting a file (in the library or simply a file), Kodi starts playing the crashes after a couple of seconds and after audio is heard for a few microseconds or so. Kodi in Windows crashes back to Windows. Kodi on Linux crashes then re-starts at the main Kodi menu.

## Possible Fix

Reading the logs it looks like there's a cache reading problem?

The the Kodi Crash log for the MKV I see this:

2019-10-01 10:50:30.682 T:140475424659200   DEBUG: CDVDInputStreamFile::SetReadRate - set cache throttle rate to 3123413 bytes per second

Is this cache throttle rate too small for 16,000kbps file?

### To Reproduce

Change Audio Settings to use HDMI Pass-through and set the AVR Surround format to TrueHD enabled

Download the two small 30 second sample files here:
https://drive.google.com/drive/folders/1Dv8rQ00Ur_6tVb1iXc1iZPCYMiT96gRp?usp=sharing

1.  Copy the sample mka to a Kodi audio source
2.  Play the file (from Music menu)
3.  Kodi crashes 

1.  Copy the sample mkv to a Kodi video source
2.  Play the file (from Video menu)
3.  Kodi crashes 

### Debuglog

The debuglogs can be found here:

LibreElec 9.2 on NUC8i3 playing MKA file:
https://paste.kodi.tv/efoqikuhax.kodi

LibreElec 9.2 on NUC8i3 playing MKV file (same audio stream)
https://paste.kodi.tv/konunolupo.kodi

Kodi 18.2 on Windows x64 palying MKA file:
https://paste.kodi.tv/qibofewuja.kodi

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [X] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [X] Windows
 - [ ] Windows UWP

 - Operating system version/name:
 - Kodi version:
18.4 latest official (LibreElec 9.2 and Windows 10 Pro x64)
18.4 CoreElec Nightly dated 1 Oct 2019

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by HomerJau** [comment]
Here's the audio codec info from mediainfo:

Audio
--
ID : | 1
Format : | MLP FBA 16-ch
Format/Info : | Meridian Lossless Packing FBA with 16-channel  presentation
Commercial name : | Dolby TrueHD with Dolby Atmos
Codec ID : | A_TRUEHD
Duration : | 30 s 73 ms
Bit rate mode : | Variable
**Bit rate : | 11.9 Mb/s
Maximum bit rate : | 16.3 Mb/s**
Channel(s) : | 8 channels
Channel layout : | L R C LFE Ls Rs Lb Rb
Sampling rate : | 48.0 kHz
**Frame rate : | 1 200.000 FPS (40 SPF)**
Compression mode : | Lossless
Stream size : | 42.7 MiB (99%)
Title : | Surround 7.1
Language : | English
Default : | No
Forced : | No
Number of dynamic objects : | 15
Bed channel count : | 1 channel
Bed channel configuration : | LFE

**Comment by KarellenX** [comment]
I was able to confirm this issue using the posted sample files.

System Win10; v18.4 passthrough enabled into Denon amp.

mkv file played via video library. Log line 1459... https://paste.kodi.tv/udimihedoz.kodi

mka file played via music library. Log line 1581... https://paste.kodi.tv/gazorimahe.kodi

**Comment by MilhouseVH** [comment]
I know this mentions Atmos in addition to TrueHD, but I wonder if the underlying issue (whatever that may be) is also the cause of https://github.com/xbmc/xbmc/issues/16096

**Comment by KarellenX** [comment]
I did wonder that also and it is quite possible, but from my limited knowledge the errors in the log seem different.

**Comment by anssih** [comment]
I created PR #16813 that should fix the crash.

Audio dropouts still remain. Those will be much harder to fix. I do have a WIP patch for ffmpeg that makes the audio work, but will still have to clean it up and get the fix in Kodi codebase as well.

**Comment by HomerJau** [comment]
Thanks for the update Anssi.

Hopefully the crash fix and the ffmpeg update to fix Atmos skipping when complete can be merged back into Kodi 18.

**Comment by danieldjewell** [comment]
I think I found a related issue -- Kodi 18.4 on Android (NV Shield TV) doesn't crash but the audio does cut out during what I presume to be high bitrate parts. 

I'll see if I can't test some more but thought it might be related to this.

**Comment by V33m** [comment]
How is the feedback over at OSMC given it has been pushed to general release? There are three samples with stuttering over at CoreELEC forum if you guys need any more samples for debugging: [TrueHD with Dolby Atmos 7.1 sound skipping (11Mbps audio bit rate)](https://discourse.coreelec.org/t/truehd-with-dolby-atmos-7-1-sound-skipping-11mbps-audio-bit-rate/7569/15)

**Comment by anssih** [comment]
The TrueHD passthrough crashes should now be fixed in Kodi master as of 2019-12-22 d29c84c9d37, and on Kodi 18.x (Leia) nightlies as of 2019-12-26 51eff914.

I'm sorry for the delays.

However, as noted previously, audio dropouts still remain, so I'll keep this issue open.
The needed changes are known but they will unfortunately take some time to implement.

**Comment by danieldjewell** [comment]
So, to add what I can: I've noticed dropouts on TrueHD/Atmos high bitrate streams only (and only during what I would assume are higher than normal levels of audio data -- I haven't researched if TrueHD/Atmos streams are ultimately a variable bitrate [i.e. during particularly "noisy" sections the amount of audio data goes up, but I'm thinking yes?]....) I initially thought that it was my A/V receiver not being able to process the stream but it's TrueHD/Atmos certified so...

**Comment by dabl2** [comment]
> However, as noted previously, audio dropouts still remain, so I'll keep this issue open.
> The needed changes are known but they will unfortunately take some time to implement.

As I understand it the needed changes that are known to fix the audio dropouts are related to ffmpeg correct? 

Might you (or anyone else) have a link to the status and/or discussion on what stage the version of ffmpeg with those changes is at?

**Comment by HomerJau** [comment]
There's info recent here:
https://forum.kodi.tv/showthread.php?tid=347674&pid=2922332#pid2922332

**Comment by anssih** [comment]
Yep, that link is good.

I have also some *very* rough work-in-progress code for FFmpeg [here](https://github.com/anssih/FFmpeg/commits/fix/truehd-wip) which seems to work, but doesn't trivially port over to Kodi as we have completely different code doing that.

**Comment by KarellenX** [comment]
Is this a related issue?
https://forum.kodi.tv/showthread.php?tid=351405

I have been able to replicate the reported problem on my setup.

**Comment by anssih** [comment]
I've just applied a fix for this in FFmpeg: https://git.ffmpeg.org/gitweb/ffmpeg.git/commitdiff/36e156bef02566d70cea46cc5e00b3e5d5ed3286

I intend to port it over to Kodi next.

@KarellenX It is not.

**Comment by anssih** [comment]
I pushed some initial work into https://github.com/anssih/xbmc/tree/truehd-wip.

It doesn't work yet (or at least I still got dropouts) - I suspect one of the "BUG" comments I added in ActiveAESink.cpp describe the cause (and they *might* be non-trivial to fix due to the assumptions in the existing code).

**Comment by dread185** [comment]
@anssih 
I suggest contacting Shield@NVIDIA via the Nvidia Shield TV forums:
https://www.nvidia.com/en-us/geforce/forums/shield-tv/9/326572/2019-shield-experience-upgrade-810-hotfix-image/

Nvidia FW Devs have recently fixed the high bitrate audio (TrueHD) dropout audio issue occurring at fixed locations. Fixed in Hotfix #6 for OS V8.10. Bug was in the Android OS open source libraries.

Plex now works flawlessly since it leverages only the Android OS libraries but Kodi still suffers from the high bitrate audio (TrueHD) dropouts.

Shield@NVIDIA stated Nvidia SW devs are willing to work with Kodi Devs to help Kodi resolve this issue as well.

Hope this info is of some help.

**Comment by fritsch** [comment]
Plex uses Android's Packer.
Shield uses kodi's IEC Packer.

And: Nvidia devs contacted us some weeks ago. Basically their fix was backporting Android 10s new object MAT audio format.

As kodi is not Android only and especially not Shield only, we fix the rootcause via ffmpeg

**Comment by dread185** [comment]
Ok. Got it.

**Comment by C4Wiz** [comment]
> I've just applied a fix for this in FFmpeg: https://git.ffmpeg.org/gitweb/ffmpeg.git/commitdiff/36e156bef02566d70cea46cc5e00b3e5d5ed3286
> 
> I intend to port it over to Kodi next.

how goes the good fight? any progress?

**Comment by MarkToon** [comment]
Is this issue still ongoing?

**Comment by HomerJau** [comment]
@MarkToon The Kodi Crash has been fixed buy high bit rate TrueHD (Atmos) still causes audio severe dropouts/skipping. There's a known fix buy no dev resources to have it fixed yet. Hopefully by Kodi 19.0?

**Comment by C4Wiz** [comment]
i will donate $50 to whom. ever gets this done in kodi v18

**Comment by C4Wiz** [comment]
i can not believe this has not been fixed yet!

Matrix is a year off, Leia is now!!! this needs to be fixed for the current release

**Comment by crisosma** [comment]
@anssih
Is your solution implemented on the new kodi 18.7?
I took a look at the change log but I could´t find anything that resembles that this problem was fixed.
Thank you very much!

**Comment by C4Wiz** [comment]
Nope

**Comment by hugepants** [comment]
A plausible description and solution here:
[mlp: A Dolby TrueHD demuxer and utility tool, with a focus on accurate and correct TrueHD demuxing. ](https://github.com/domyd/mlp)

**Comment by fritsch** [comment]
Great. Can you send a Pull-Request?

**Comment by hugepants** [comment]
@fritch No need, as the core issue stems from inaccurate rips of TrueHD tracks. It will be solved when ripping/demuxer software implement the changes and users' content is subsequently re-ripped.

**Comment by fritsch** [comment]
I don't agree, as other players can workaround it, kodi should do as well. But as you know: time is limited, therefore we are thankful for an outside contribution.

**Comment by HomerJau** [comment]
I can confirm the mlp Dolby TrueHD demuxer DOES NOT solve the high bitrate skipping issue in Kodi 18.6 with my: The Beatles - Abbey Road (Atmos BDA). Exact same issue as always.

I re-ripped the BDA, ran mlp tool on the playlist. It successfully extracted the TrueHD audio. I remuxed into a new MKA file. Played in Kodi (skip, skip, skip).

**Comment by domyd** [comment]
Author of the `mlp` tool here. The thing I wanted to fix with that were rips that weren't perfectly A/V synced and had noticeable audio issues at segment boundaries. I've never encountered TrueHD dropout issues in any player on any disc/movie, so that wasn't on my radar. I'm not entirely convinced this can be solved on the demuxer's part.

Do these dropouts ever happen locally or only over HDMI? Does anyone have those when playing back the original blu-ray `.m2ts` files?

**Comment by fritsch** [comment]
Hi,

thanks for posting here. I think the issues are quite mixed now. The initial issue that causes the dropouts are concerning the framesize of TrueHD and how equally distributed they are assumed. Some guys interpret the spec differently and state that the max payload size is an average size, means: sometimes it can be a bit larger, while next time a bit lower.

The fix here makes that clear: https://git.ffmpeg.org/gitweb/ffmpeg.git/commitdiff/36e156bef02566d70cea46cc5e00b3e5d5ed3286

Kodi historically uses its own syncers and parsers. Here TrueHD is assumed "static" as well.

On the long-term we plan to remove this duplicated code and rely more on ffmpeg's spdif interface. But yeah real life is more demanding nowadays, more than it was back at university.

**Comment by budtz** [comment]
Any news on this? Is kodi just not gonna work with disney Blu-ray anymore? Very sad but I guess there isnt the time.

**Comment by Krobar** [comment]
Also interested in this. Am I right in sating this is too big of a change for a patch release and not currently scheduled for Matrix either?

**Comment by fritsch** [comment]
No, you are wrong. The fix can go in every time, but as of now no one has the time working on it and as we have seen it's not trivial for the way kodi implements packing and syncing of PT, which is currently fixed size for truehd.

**Comment by dshokouhi** [comment]
Is it a huge change to do if we wanted to do a quick and dirty fix for the shield only?  I know Kodi wants to support all platforms but if we can help a handful of users that would be really great.  If anything just as a proof of concept that it can be fixed in one place.  I don't know much of the Kodi architecture but thought I would ask.  It does seem like a big change but wanted to ask the experts :)

**Comment by fritsch** [comment]
If it's done for the shield it's done for the rest as well, as they share the code.

But as you talk about the WE, go ahead and let's see what WE can come up with.

**Comment by bendavid** [comment]
@anssih I was going to take a look at this, but actually your commit ( https://github.com/anssih/xbmc/commit/2a21581c1afd9045a64b5cb2b98a8d6b6bd5c989 ) already fixes the audio drops in the clips I was testing.  Can you point me to some content which still had issues?

**Comment by hugepants** [comment]
[This discussion](https://discourse.osmc.tv/t/atmos-issues/72920/164) mentions some example scenes in Frozen 2 and The Incredibles. I believe several Disney movies are affected.

**Comment by depechefan** [comment]
If you are able to find a sample from The Beatles, Abbey Road in Dolby Atmos then that is a great way to test. Let me know if you need me to put something together for you to test with.

**Comment by bendavid** [comment]
Ok, so https://github.com/anssih/xbmc/commit/2a21581c1afd9045a64b5cb2b98a8d6b6bd5c989 makes the situation SIGNIFICANTLY better.  There are still occasional skips in Abbey Road, and I confirmed by counting invocations of ```CAEPackIEC61937::PackTrueHD``` that the two "BUG" comments at https://github.com/anssih/xbmc/commit/2a21581c1afd9045a64b5cb2b98a8d6b6bd5c989#diff-05e1b56df10155e0212d39b1d73fe66aa0712448de99811146e47688c9614b3bR944 and https://github.com/anssih/xbmc/commit/2a21581c1afd9045a64b5cb2b98a8d6b6bd5c989#diff-05e1b56df10155e0212d39b1d73fe66aa0712448de99811146e47688c9614b3bR952 are related in that both of those issues appear right around where the skipping occurs.  (e.g. cases with 2 or 0 invocations of   ```CAEPackIEC61937::PackTrueHD``` adjacent to each other.)

I'll need to understand the code a bit better before I can fix it properly.

If it takes a while it's probably worth already taking the existing commit for the moment since that will likely raise the bitrate threshold at which problems occur.

**Comment by HomerJau** [comment]
Apologies. I had removed the original samples with issues. The link in my OP is now changed: 

https://drive.google.com/drive/folders/1E1EGkP-Kv5IbPufU1ByMFHBf0MYqaCv0?usp=sharing

30 sec samples: MKV and MKA file with Atmos at Maximum bit rate : | 16.3 Mb/s:

**Comment by dabl2** [comment]
Bless you anssih and bendavid for picking up the flag on this.  Would love to see this bug get squashed.

**Comment by apguy** [comment]
Yes.  I concur.   Thank you for looking at this

On 19 May 2021, at 7:39 am, Don Lindbergh ***@***.***> wrote:

﻿

Bless you anssih and bendavid for picking up the flag on this. Would love to see this bug get squashed.

**Comment by Deamaed** [comment]
> Ok, so [anssih@2a21581](https://github.com/anssih/xbmc/commit/2a21581c1afd9045a64b5cb2b98a8d6b6bd5c989) makes the situation SIGNIFICANTLY better. There are still occasional skips in Abbey Road, and I confirmed by counting invocations of `CAEPackIEC61937::PackTrueHD` that the two "BUG" comments at [anssih@2a21581#diff-05e1b56df10155e0212d39b1d73fe66aa0712448de99811146e47688c9614b3bR944](https://github.com/anssih/xbmc/commit/2a21581c1afd9045a64b5cb2b98a8d6b6bd5c989#diff-05e1b56df10155e0212d39b1d73fe66aa0712448de99811146e47688c9614b3bR944) and [anssih@2a21581#diff-05e1b56df10155e0212d39b1d73fe66aa0712448de99811146e47688c9614b3bR952](https://github.com/anssih/xbmc/commit/2a21581c1afd9045a64b5cb2b98a8d6b6bd5c989#diff-05e1b56df10155e0212d39b1d73fe66aa0712448de99811146e47688c9614b3bR952) are related in that both of those issues appear right around where the skipping occurs. (e.g. cases with 2 or 0 invocations of `CAEPackIEC61937::PackTrueHD` adjacent to each other.)
> 
> I'll need to understand the code a bit better before I can fix it properly.
> 
> If it takes a while it's probably worth already taking the existing commit for the moment since that will likely raise the bitrate threshold at which problems occur.

Hi there.  I am not a coder, but am using 17.7 DS Player with MadVR (and external LAV filters) to get everything working properly, and it's finicky.  I'd prefer to use Kodi 19.

If there is a way to test this, including in Kodi 19 - I'd be happy to test it.  I have several of the usual suspects to test, for example, Frozen 2, Zootopia, that have repeatable dropouts at set spots.  

Let me know.

**Comment by apollolucas** [comment]
@anssih @bendavid Any update on this bug? How about bendavid’s suggestion to use the existing commits of anssih to remedy the situation somewhat? As more and more BD movies use higher bitrates this will become a major problem for Kodi users. I’ve been experiencing the bug on many more of my ripped movies lately (not limited to Disney anymore).
I really appreciate your efforts and help!

**Comment by Deamaed** [comment]
> @anssih @bendavid Any update on this bug? How about bendavid’s suggestion to use the existing commits of anssih to remedy the situation somewhat? As more and more BD movies use higher bitrates this will become a major problem for Kodi users. I’ve been experiencing the bug on many more of my ripped movies lately (not limited to Disney anymore).
> I really appreciate your efforts and help!

In my post above - I was just trying to get guidance on how you can actually integrate ansshi's commits into the current Kodi to test?  It appears bendavid did this - I just don't know enough to do this, but would be willing to test as I have several movies that I know the spots that skip and would be able to test the impact.

**Comment by Deamaed** [comment]
Just as a follow-up, I decided to try an amateur job of making a new build of Kodi using the changes proposed by anssih.  I was using the current version of Kodi and that version would cause no audio, or the video to freeze.  Perhaps I made an error in making the changes or just in creating the exe.

I'm curious if @bendavid incorporated this code into an older version of Kodi, or the newest one.  I'd be happy with even a half reduction in the dropouts.

**Comment by apguy** [comment]
Still having this problem. Normally it's tolerable. One or two dropouts per movie. 

Today I watched free guy.  So many dropouts it was unwatchable.  Had to turn off bitstreaming and therefore listen to straight 7.1 instead of atmos

**Comment by R0000** [comment]
> Today I watched free guy. So many dropouts it was unwatchable. Had to turn off bitstreaming and therefore listen to straight 7.1 instead of atmos

Yes I had the same issue with Free Guy (on an nvidia shield). However VLC on it actually worked well (had judder in a couple of scenes though). My version was Dolby Digital+ 7.1.

**Comment by C4Wiz** [comment]
yep, this siiue is very old and getting worse! the bad news is KODI drvs just don't care about such a core feature. new versions with the same issue is more importaint!

it started in 18, still present in 19, and guess what......still present in 20.

what a joke......my advise is to move to a player with a native player that does not have this issue. look at zidoo

**Comment by apollolucas** [comment]
Free Guy is indeed unwatchable with Atmos. Been noticing this problem more and more lately with newer movies. Kind of starts to make Kodi obsolete now.
Hoping that if more people post here the Devs will show mercy and pick it up again.
As others have suggested I’d be willing to donate to get this problem fixed.

**Comment by lrusak** [comment]
+1's and me too's don't help here so I'm going to lock the thread to devs only.

No one is actively working on this from Team Kodi that I'm aware of so I'll leave it open for now.

For the non devs please take your conversation to the forum. Thanks.
